### PR TITLE
Fail when stating unreadable files

### PR DESCRIPTION
**Pull request by wolfs**
We had some problems masked by treating files we could not stat as missing, see for example #17551. With the new `@Untracked` annotation we also give folks a tool how to deal with such files in their custom tasks. So let's fail when we are unable to `stat` a file and not silently return a stat for a missing file.

### CODE REVIEW

File: subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/jdk7/NioFileMetadataAccessor.java [hunk 1]
```diff
@@ -50,7 +51,7 @@ public FileMetadata stat(File file) {
             return DefaultFileMetadata.directory(accessType);
         }
         if (attributes.isOther()) {
-            return DefaultFileMetadata.missing(accessType);
+            throw new UncheckedIOException("Unsupported file type for " + file.getAbsolutePath());
```
Comment (lptr) [inline subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/jdk7/NioFileMetadataAccessor.java:54]: Can we log something here about the actual type?

File: subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/services/NativePlatformBackedFileMetadataAccessor.java [hunk 2]
```diff
@@ -39,14 +40,18 @@ public FileMetadata stat(File f) {
...
         AccessType accessType = AccessType.viaSymlink(stat.getType() == FileInfo.Type.Symlink);
         if (accessType == AccessType.VIA_SYMLINK) {
             try {
                 stat = files.stat(f, true);
             } catch (NativeException e) {
-                return DefaultFileMetadata.missing(AccessType.VIA_SYMLINK);
+                // For a symlink cycle, file.exists() returns false when unable to stat the file.
+                if (!f.exists()) {
+                    return DefaultFileMetadata.missing(accessType);
+                }
```
Comment (lptr) [inline subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/services/NativePlatformBackedFileMetadataAccessor.java:53]: So we'll still represent a symlink cycle as a missing file?

**Review by lptr — COMMENTED** [review]
Looks good in general, I left a couple comments. What surprises me somewhat is that we need to add this behavior in so many places.

File: subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/jdk7/NioFileMetadataAccessor.java (hunk 1 above)
Comment (wolfs) [reply subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/jdk7/NioFileMetadataAccessor.java:54]: The type is "Other", so it doesn't make much sense to log it. I used to log it first and then realized that it doesn't make much sense.

File: subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/services/NativePlatformBackedFileMetadataAccessor.java (hunk 2 above)
Comment (wolfs) [reply subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/services/NativePlatformBackedFileMetadataAccessor.java:53]: Yes. That is what happens with all but the native accessors, so I'd do this for the native implementation as well. It is then represented the same way as a broken symlink, which I think is fine.

**Comment by wolfs** [comment]
> What surprises me somewhat is that we need to add this behavior in so many places.

The problem here is that we have multiple implementations for stating files. Depending on the Java version on whether or not native platform is available, we use one of them.Since we use these services in workers as well, we need to support Java version 6.

File: subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/services/NativePlatformBackedFileMetadataAccessor.java (hunk 2 above)
Comment (lptr) [reply subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/services/NativePlatformBackedFileMetadataAccessor.java:53]: But if you tried to actually open that file, you'd get a `FileSystemLoopException`, no? To me a cycling symlink sounds very much like unreadable content, not a missing file.

Comment (wolfs) [reply subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/services/NativePlatformBackedFileMetadataAccessor.java:53]: Good question. When we walk a directory, we currently ignore file system loop exceptions: https://github.com/gradle/gradle/blob/bfcb154efdc7622a51352a37b284a78888857eb8/subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java#L356-L365

Comment (lptr) [reply subprojects/native/src/main/java/org/gradle/internal/nativeintegration/filesystem/services/NativePlatformBackedFileMetadataAccessor.java:53]: Alright. I guess your proposal doesn't change behavior, so it's fine if we stay a bit inconsistent for this rare use case.

**Review by lptr — APPROVED** [review]
LGTM.

**Comment by wolfs** [comment]
@bot-gradle test and merge

**Review by lptr — APPROVED** [review]
LGTM.

**Pull request merged**